Bottom: 30b3c74e18a864b8392349a6407b343a29bfdddd
Top:    3db371792cd682d85f634f62a487c2a95b42ee6f
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-09-13 17:10:09 -0500

Configure an HAProxy reverse proxy

---
diff --git a/ansible-playbook/group_vars/all/credentials.yml b/ansible-playbook/group_vars/all/credentials.yml
index e7a2255bc..921381ed6 100644
--- a/ansible-playbook/group_vars/all/credentials.yml
+++ b/ansible-playbook/group_vars/all/credentials.yml
@@ -17,6 +17,7 @@ gitlab_db_password:                  '{{ lookup("password", playbook_dir + "/cre
 gitlab_email_password:               '{{ lookup("password", playbook_dir + "/credentials/git/email/gitlab@novatech-llc.com"   , length=20) }}'
 gitlab_root_password:                '{{ lookup("password", playbook_dir + "/credentials/git/users/root"                      , length=20) }}'
 gitlab_secrets_db_key_base:          '{{ lookup("password", playbook_dir + "/credentials/git/key/GITLAB_SECRETS_DB_KEY_BASE"  , length=40) }}'
+haproxy_stats_password:              '{{ lookup("password", playbook_dir + "/credentials/haproxy/stats"                       , length=20) }}'
 mantisbt_db_password:                '{{ lookup("password", playbook_dir + "/credentials/mantis/db/user"                      , length=20) }}'
 mantisbt_db_root_password:           '{{ lookup("password", playbook_dir + "/credentials/mantis/db/root"                      , length=20) }}'
 mantisbt_email_password:             '{{ lookup("password", playbook_dir + "/credentials/mantis/email/mantis@novatech-llc.com", length=20) }}'
diff --git a/ansible-playbook/group_vars/docker-hosts/networking.yml b/ansible-playbook/group_vars/docker-hosts/networking.yml
index f652891ad..38e7fdc6b 100644
--- a/ansible-playbook/group_vars/docker-hosts/networking.yml
+++ b/ansible-playbook/group_vars/docker-hosts/networking.yml
@@ -1,30 +1,20 @@
 ---
 
 svn_hostname: svn.novatech-llc.com
-svn_port_args:
-- "{{ svn_ip_addr }}:80:80"
-- "{{ svn_ip_addr }}:443:443"
+svn_port_args: []
 
 gitlab_hostname: git.novatech-llc.com
 gitlab_port_args:
-- "{{ gitlab_ip_addr }}:80:80"
-- "{{ gitlab_ip_addr }}:443:443"
 - "{{ gitlab_ip_addr }}:22:22"
 
 wiki_hostname: wiki.novatech-llc.com
-wiki_port_args:
-- "{{ wiki_ip_addr }}:80:80"
-- "{{ wiki_ip_addr }}:443:443"
+wiki_port_args: []
 
 mantisbt_hostname: mantis.novatech-llc.com
-mantisbt_port_args:
-- "{{ mantisbt_ip_addr }}:80:80"
-- "{{ mantisbt_ip_addr }}:443:443"
+mantisbt_port_args: []
 
 build_hostname: buildsystem.novatech-llc.com
-build_port_args:
-- "{{ build_ip_addr }}:80:80"
-- "{{ build_ip_addr }}:443:443"
+build_port_args: []
 
 testssh_hostname: testssh.novatech-llc.com
 testssh_port_args:
@@ -42,19 +32,22 @@ openldap_port_args:
 - "{{ openldap_ip_addr }}:636:636"
 
 phpldapadmin_hostname: ldap.novatech-llc.com
-phpldapadmin_port_args:
-- "{{ phpldapadmin_ip_addr }}:80:80"
-- "{{ phpldapadmin_ip_addr }}:443:443"
+phpldapadmin_port_args: []
 
 phpmyadmin_hostname: phpmyadmin.novatech-llc.com
 
 buildbot_hostname: buildbot.novatech-llc.com
 buildbot_port_args:
-- "{{ buildbot_ip_addr }}:80:8080"
-- "{{ buildbot_ip_addr }}:443:8443"
 - "{{ buildbot_ip_addr }}:9989:9989"
 
 nexus_hostname: nexus.novatech-llc.com
-nexus_port_args:
-- "{{ nexus_ip_addr }}:80:8081"
-- "{{ nexus_ip_addr }}:443:8443"
+nexus_port_args: []
+
+haproxy_port_args:
+- "{{ ansible_facts.default_ipv4.address }}:80:80"
+- "{{ ansible_facts.default_ipv4.address }}:443:443"
+- "{{ ansible_facts.default_ipv4.address }}:9000:9000"
+- "{{ buildbot_ip_addr }}:80:80"
+- "{{ buildbot_ip_addr }}:443:443"
+- "{{ nexus_ip_addr }}:80:80"
+- "{{ nexus_ip_addr }}:443:443"
diff --git a/ansible-playbook/site.yml b/ansible-playbook/site.yml
index b4ff3af83..fcdb207fb 100644
--- a/ansible-playbook/site.yml
+++ b/ansible-playbook/site.yml
@@ -22,7 +22,7 @@
   become_user: root
   become_method: sudo
   tags: daedalus
-  roles: 
+  roles:
   - role: daedalus
 
 - name: Bacula
@@ -87,6 +87,10 @@
     role: docker-openssl
     tags:
     - openssl_container
+  - name: Build and start haproxy reverse proxy container
+    role: haproxy
+    tags:
+    - haproxy_container
   - name: Build and start Nexus Artifact Repository container
     role: nexus3
     tags:
diff --git a/portmap.txt b/portmap.txt
new file mode 100644
index 000000000..6318bf53e
--- /dev/null
+++ b/portmap.txt
@@ -0,0 +1,118 @@
+              *:111
+              *:9101
+              *:9103
+             :::111
+            ::1:25
+127.000.000.001:25      smtp
+127.000.000.001:9102    bacula-fd
+
+172.016.000.100:22      ssh
+172.016.000.100:3306    mysql
+
+svn.novatech-llc.com
+172.016.000.101:443     https
+172.016.000.101:80      http
+
+git.novatech-llc.com
+172.016.000.102:22      ssh
+172.016.000.102:443     https
+172.016.000.102:80      http
+
+wiki.novatech-llc.com
+172.016.000.103:443     https
+172.016.000.103:80      http
+
+mantis.novatech-llc.com
+172.016.000.104:443     https
+172.016.000.104:80      http
+
+buildsystem.novatech-llc.com
+172.016.000.105:22      ssh
+172.016.000.105:3306    mysql
+172.016.000.105:80      http
+
+testssh.novatech-llc.com
+172.016.000.106:22      ssh
+
+mail.novatech-llc.com
+172.016.000.107:25      smtp
+172.016.000.107:465     smtps
+172.016.000.107:587     smtp+starttls
+
+ldap.novatech-llc.com
+172.016.000.108:389     ldap
+172.016.000.108:443     https
+172.016.000.108:636     ldaps
+172.016.000.108:80      http
+
+buildbot.novatech-llc.com
+172.016.000.110:443     https
+172.016.000.110:80      http
+172.016.000.110:9989    buildbot
+
+nexus.novatech-llc.com
+172.016.000.111:443     https
+172.016.000.111:80      http
+
+frankensteinbs.novatech-llc.com
+172.016.000.120:22      ssh
+172.016.000.120:3306    mysql
+172.016.000.120:69      tftp
+172.016.000.120:80      http
+
+192.168.122.001:53      dns
+
+ssh:22
+    daedalus
+    git
+    buildsystem
+    testssh
+    frankensteinbs
+
+smtp:25
+    mail
+
+tftp:69
+    frankensteinbs
+
+http:80
+    svn
+    git
+    wiki
+    mantis
+    buildsystem
+    ldap
+    buildbot
+    nexus
+    frankensteinbs
+
+ldap:389
+    ldap
+
+https:443
+    svn
+    git
+    wiki
+    mantis
+    buildsystem
+    ldap
+    buildbot
+    nexus
+    frankensteinbs
+
+smtps:465
+    mail
+
+smtp+starttls:587
+    mail
+
+ldaps:636
+    ldap
+
+mysql:3306
+    daedalus
+    buildsystem
+    frankensteinbs
+
+buildbot:9989
+    buildbot
diff --git a/roles/haproxy/defaults/main.yml b/roles/haproxy/defaults/main.yml
new file mode 100644
index 000000000..5951965ba
--- /dev/null
+++ b/roles/haproxy/defaults/main.yml
@@ -0,0 +1,32 @@
+---
+
+haproxy_port_args: []
+
+image_name: "haproxy:1.8.12-alpine"
+
+config_cert_path: /etc/ssl/private
+
+# host path containing configuration files
+config_hostdir: "{{ docker_restore_config_base_dir }}/{{ container_name }}"
+
+# environment passed to the container
+container_env: {}
+
+# hostname
+container_hostname: "{{ container_name }}"
+
+# the name of the container being started
+container_name: "{{ docker_name_prefix }}haproxy"
+
+# networks of the container
+container_networks:
+  - name: '{{ docker_network_frontend }}'
+    aliases:
+      - '{{ container_hostname }}'
+
+# exposed ports of the container
+container_port_args: '{{ haproxy_port_args }}'
+
+container_volumes:
+  - "{{ config_hostdir }}:/usr/local/etc/haproxy:ro"
+  - "{{ openssl_dv_name }}:{{ config_cert_path }}:z"
diff --git a/roles/haproxy/tasks/main.yml b/roles/haproxy/tasks/main.yml
new file mode 100644
index 000000000..533ab9c3d
--- /dev/null
+++ b/roles/haproxy/tasks/main.yml
@@ -0,0 +1,35 @@
+---
+# This will generate passwords for these accounts.
+- assert:
+    that:
+    - haproxy_stats_password is defined
+
+# Setup the configuration directory
+- name: docker_container.conf dir
+  file:
+    state: directory
+    path: '{{ config_hostdir }}'
+    owner: root
+    group: root
+    mode: 'u=rwx,g=rx,o=rx'
+    recurse: no
+
+
+- name: Populate config files
+  template:
+    src: "{{ item }}.j2"
+    dest: "{{ config_hostdir }}/{{ item }}"
+  loop:
+    - haproxy.cfg
+
+- name: Start haproxy container
+  docker_container:
+    name: '{{ container_name }}'
+    env: '{{ container_env }}'
+    hostname: '{{ container_hostname }}'
+    image: '{{ image_name }}'
+    networks: '{{ container_networks }}'
+    ports: '{{ container_port_args }}'
+    purge_networks: true
+    state: started
+    volumes: '{{ container_volumes }}'
diff --git a/roles/haproxy/templates/haproxy.cfg.j2 b/roles/haproxy/templates/haproxy.cfg.j2
new file mode 100644
index 000000000..228ef73a0
--- /dev/null
+++ b/roles/haproxy/templates/haproxy.cfg.j2
@@ -0,0 +1,93 @@
+global
+    daemon
+    maxconn 256
+    ssl-dh-param-file {{ config_cert_path }}/dhparam.pem
+
+defaults
+    mode http
+    timeout connect 5s
+    timeout client 50s
+    timeout server 50s
+    default-server init-addr last,libc,none
+
+#enable stats page
+listen stats
+    bind *:9000
+    mode http
+    stats enable
+
+#
+frontend container_https_in
+    mode http
+    bind *:80
+    bind *:443 ssl crt {{ config_cert_path }}/haproxy.pem
+
+    http-request set-header X-Forwarded-Proto https
+    http-request set-header X-SSL %[ssl_fc]
+
+    http-request redirect scheme https code 301 if !{ ssl_fc }
+
+    acl is_buildbot     hdr(host) eq {{ buildbot_hostname }}
+    acl is_gitlab       hdr(host) eq {{ gitlab_hostname }}
+    acl is_ldapadmin    hdr(host) eq {{ openldap_hostname }}
+    acl is_mantisbt     hdr(host) eq {{ mantisbt_hostname }}
+    acl is_nexus        hdr(host) eq {{ nexus_hostname }}
+    acl is_svn          hdr(host) eq {{ svn_hostname }}
+    acl is_wiki         hdr(host) eq {{ wiki_hostname }}
+
+    use_backend backend-buildbot if is_buildbot
+    use_backend backend-nexus if is_nexus
+    use_backend backend-svn if is_svn
+    use_backend backend-ldapadmin if is_ldapadmin
+    use_backend backend-gitlab if is_gitlab
+    use_backend backend-mantisbt if is_mantisbt
+    use_backend backend-wiki if is_wiki
+
+# frontend container_ldap_in
+#     mode tcp
+#     bind *:389
+#     bind *:636
+#     use_backend backend-ldap
+
+# frontend container_smtp_in
+#     mode tcp
+#     bind *:25
+#     bind *:465
+#     bind *:587
+#     use_backend backend-exim4
+
+# frontend container_ssh_in
+#     mode tcp
+#     bind *:22
+#     default_backend
+
+#
+backend backend-svn
+    server svn1 svn.frontend:80
+
+backend backend-ldapadmin
+    server phpldapadmin1 phpldapadmin.frontend:80
+
+backend backend-exim4
+    server exim4_1 exim4.frontend:25
+
+backend backend-gitlab
+    server gitlab1 gitlab.frontend:80
+
+backend backend-wiki
+    server wiki1 wiki.frontend:80
+
+backend backend-mantisbt
+    server mantisbt1 mantisbt.frontend:80
+
+backend backend-build
+    server build1 build.frontend:80
+
+backend backend-testssh
+    server testssh1 testssh.frontend:80
+
+backend backend-buildbot
+    server buildbot1 buildbot.frontend:8080
+
+backend backend-nexus
+    server nexus1 nexus3.frontend:8081
